flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptivebatch/AdaptiveBatchSchedulerTest.java

```diff
@@ -0,0 +1,180 @@
+            ComponentMainThreadExecutorServiceAdapter.forMainThread();
+
+    @Test
+    public void testAdaptiveBatchScheduler() throws Exception {
+        final JobVertex source1 = createJobVertex("source1", SOURCE_PARALLELISM_1);
+        final JobVertex source2 = createJobVertex("source2", SOURCE_PARALLELISM_2);
+        final JobVertex sink = createJobVertex("sink", -1);
+        sink.connectNewDataSetAsInput(
+                source1, DistributionPattern.POINTWISE, ResultPartitionType.BLOCKING);
+        sink.connectNewDataSetAsInput(
+                source2, DistributionPattern.POINTWISE, ResultPartitionType.BLOCKING);
+
+        Configuration configuration = new Configuration();
+        configuration.set(
+                JobManagerOptions.SCHEDULER, JobManagerOptions.SchedulerType.AdaptiveBatch);
+
+        final JobGraph jobGraph = new JobGraph(new JobID(), "test job", source1, source2, sink);
+
+        final AdaptiveBatchSchedulerTestUtils.AdaptiveBatchSchedulerBuilder schedulerBuilder =
+                (AdaptiveBatchSchedulerTestUtils.AdaptiveBatchSchedulerBuilder)
+                        new AdaptiveBatchSchedulerTestUtils.AdaptiveBatchSchedulerBuilder(
+                                        jobGraph, mainThreadExecutor)
+                                .setJobMasterConfiguration(configuration);
+        schedulerBuilder.setJobVertexParallelismDecider((ignored) -> 10);
+
+        final AdaptiveBatchScheduler scheduler = schedulerBuilder.build();
+        final DefaultExecutionGraph graph = (DefaultExecutionGraph) scheduler.getExecutionGraph();
+        final ExecutionJobVertex sinkExecutionJobVertex = graph.getJobVertex(sink.getID());
+
+        scheduler.startScheduling();
+        assertThat(sinkExecutionJobVertex.getParallelism(), is(-1));
+
+        // trigger source1 finished.
+        transitionExecutionsState(scheduler, ExecutionState.FINISHED, source1);
+        assertThat(sinkExecutionJobVertex.getParallelism(), is(-1));
+
+        // trigger source2 finished.
+        transitionExecutionsState(scheduler, ExecutionState.FINISHED, source2);
+        assertThat(sinkExecutionJobVertex.getParallelism(), is(10));
+
+        // check that the jobGraph is updated
+        assertThat(sink.getParallelism(), is(10));
+    }
+
+    @Test
+    public void testDecideParallelismForForwardTarget() throws Exception {
+        final JobVertex source1 = createJobVertex("source1", SOURCE_PARALLELISM_1);
+        final JobVertex source2 = createJobVertex("source2", SOURCE_PARALLELISM_2);
+        final JobVertex sink = createJobVertex("sink", -1);
+        sink.connectNewDataSetAsInput(
+                source1, DistributionPattern.POINTWISE, ResultPartitionType.BLOCKING);
+        sink.connectNewDataSetAsInput(
+                source2, DistributionPattern.ALL_TO_ALL, ResultPartitionType.BLOCKING);
+        source1.getProducedDataSets().get(0).getConsumer().setForward(true);
```

Comment:
The scheduler creation of the two tests are almost the same except for these 2 lines, can we extract them into two methods `createJobGraph(boolean withForwardEdge)` and `createScheduler(JobGraph)` to reduce duplication?